docs/source/format/CanonicalExtensions.rst

```diff
@@ -251,6 +251,27 @@ Variable shape tensor
    Values inside each **data** tensor element are stored in row-major/C-contiguous
    order according to the corresponding **shape**.
 
+.. _json_extension:
+
+JSON
+====
+
+* Extension name: `arrow.json`.
+
+* The storage type of this extension is ``StringArray`` or
+  or ``LargeStringArray`` or ``StringViewArray``.
+  Only UTF-8 encoded JSON is supported.
```

Comment:
I think the one thing to consider here as possible future extensions (this might have already been discussed on ML a while ago), is non-standard encodings using Binary types instead of String.

In the short term we should probably recommend that implementations should likely reject Binary annotated JSON fields

Reply:
I would suggest that binary encodings of JSON should be advertized as a different extension, e.g. "arrow.bson"

Reply:
Sorry to clarify I was not talking about BSON, but rather UTF-16 or UTF-32 encodings for JSON (which one can debate if it is worth supporting based on RFC working about requiring UTF-8).

Reply:
> By this you mean serializing binary data into UTF-8 string and including it as a String? What would be a downside to allowing that?

I think the argument was that we might want to support `BinaryArray` as a storage type since a JSON document could be encoded with a non-utf8 encoding and thus should not be stored in a `StringArray`.  However, I agree we probably don't want to worry about this since the RFC is pretty clear that JSON must be UTF-8 and users can always make up their own extension type if they need to.